### PR TITLE
ExUnit: Doctest setup + Fix bug when compiling module fails

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/ex_unit/lib/ex_unit/doc_test.ex [hunk 1]
```diff
@@ -167,12 +167,26 @@ defmodule ExUnit.DocTest do
     * `:import` — when `true`, one can test a function defined in the module
       without referring to the module name. However, this is not feasible when
       there is a clash with a module like Kernel. In these cases, `import`
-      should be set to `false` and a full `M.f` construct should be used.
+      should be set to `false` and a full `M.f` construct should be used, or
+      the `:setup` option can be used to fine-tune the configuration.
+
+    * `:setup` — takes a quoted expression of the configuration desired
+      to run for each test.
```
Comment (ericmj) [inline lib/ex_unit/lib/ex_unit/doc_test.ex]: Can we change the wording a bit here? We are not configuring anything, rather we are setting up the environment for the doctests to run with the help of the quoted expression.

**Comment by josevalim** [comment]
I don't understand what this does. Sorry.

Also notice all doctests are tagged, which means you can provide a setup for them as:

```elixir
setup context do
  if context[:doctest] do ... end
end
```

So I definitely don't want to support a special `setup` option in doctest. Specially considering that, if your doctests rely on setup, it means the code snippet you have in your documentation won't work on IEx anyway (which is the main purpose of doctests).

**Comment by eksperimental** [comment]
@josevalim yes, it does exactly what you can do with setup.

Maybe I was misled but what i read in the [description for the import option](https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/lib/ex_unit/doc_test.ex#L167-L170). 

So can I avoid a function collision with?
```elixir
setup context do
  if context[:doctest] do
    import Kernel, except: [raise: 1]
    import ExUnit.DocTestTest.ExtendedModule, only: [raise: 1]
  end
end
```

**Comment by eksperimental** [comment]
@josevalim The case for having this config, is because I am extending elixir, and I want to import certain functions with a different module, as if they were part of an elixir module.

alias is not working when used in the case you mentioned.

```elixir
  setup context do
    if context[:doctest] do
      alias ExUnit.DocTestTest.ExtendedModule, as: Extended
    end
    :ok
  end
doctest ExUnit.DocTestTest.WithSetup, import: true
```
I get
```
Doctest failed: got UndefinedFunctionError with message undefined function: Extended.foo/0 (module Extended is not available)
code:  Extended.foo()
```

**Comment by josevalim** [comment]
@eksperimental I see what you are trying to do. You can just do it before the doctest:

```elixir
alias ExUnit.DocTestTest.ExtendedModule, as: Extended
doctest ExUnit.DocTestTest.WithSetup, import: true
```

**Comment by eksperimental** [comment]
From what I can see doctest does not have access to setup, right?

Also, if we are not implementing setup, i think what we are missing is the possibility to set tags from a doctest run. maybe add the tags options.
but that wouldn't make sense we cannot acces the setup

**Comment by eksperimental** [comment]
@josevalim please also tell me what you think about the bug fix, i can submit it as a new PR. (i will remove the commentted line i forgot there)

**Comment by josevalim** [comment]
It is not about access the doctest in setup. Aliases and imports are lexical, which means that when you write this code:

```elixir
  setup context do
    if context[:doctest] do
      alias ExUnit.DocTestTest.ExtendedModule, as: Extended
      # The alias is available here
    end
    # But not here
    :ok
  end
# Nor here
```

The code you wrote wouldn't work work even with regular tests. If you want an alias in the doctests, you should just define it before the doctest.

About the bug fix, I couldn't spot it between the other changes, can you please point me to it?

**Comment by eksperimental** [comment]
@josevalim the bugfix is a commit on its own: https://github.com/eksperimental/elixir/commit/9e0690c9f6b48b106eec9962dfe59c8709243399
but i'm not sure if that's the best way to deal with it, or if the compilation fails, it should be detected before.

thanks for the explanation about the scope, I will try to understand how to do it then.

**Comment by josevalim** [comment]
@eksperimental I am not sure the fix is correct. It seems something else is going on. Why on_definition is being called if the module was not defined? Can you provide a test that reproduces the error?

**Comment by eksperimental** [comment]
https://github.com/eksperimental/elixir_fail/blob/master/test/doctest_test.exs#L58-L66
when Import ExUnit.DocTest, the test fails

```elixir
  test "fail: function clash 2" do
    assert_raise CompileError, ~r"function hd/1 imported from both ExUnit.DocTestTest.FunctionClash and Kernel, call is ambiguous", fn ->
      defmodule FunctionClashFail2 do
        import ExUnit.DocTest
        #import Kernel, except: [hd: 1]
        doctest ExUnit.DocTestTest.FunctionClash, import: true
      end
    end
```

```
1) test fail: function clash 2 (ElixirFail.DoctestTest)
     test/doctest_test.exs:58
     Expected exception CompileError but got ArgumentError (argument error)
     stacktrace:
       :erlang.++(:doctest, nil)
       (ex_unit) lib/ex_unit/case.ex:285: ExUnit.Case.__on_definition__/3
       test/doctest_test.exs:63: anonymous fn/3 in :elixir_compiler_6.__MODULE__/1
       (elixir) lib/enum.ex:1385: Enum."-reduce/3-lists^foldl/2-0-"/3
       test/doctest_test.exs:63: (module)
       (elixir) src/elixir_compiler.erl:126: :elixir_compiler.dispatch_loaded/6
       (elixir) src/elixir_module.erl:168: :elixir_module.eval_form/6
       (elixir) src/elixir_module.erl:57: :elixir_module.do_compile/5
```

the one above does not.
https://github.com/eksperimental/elixir_fail/blob/master/test/doctest_test.exs#L43-L56
and they are the same test

**Comment by josevalim** [comment]
That's because you haven't called use ExUnit.Case before, no? So it is a
bug, I am not sure the best fix though is to silently fix it.

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eksperimental** [comment]
exactly, it's when I import DocTest directly,

**Comment by josevalim** [comment]
It doesn't make sense to use it without ExUnit.Case though, does it?

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eksperimental** [comment]
well, one  of the tests in the script did it.
I created an issue for it, since this is a PR... feel free to close it if you considereing it not worth fixing it,
it just made me waster a lot of time trying to figure out why i was getting an ArgumentError when it should be a CompileError,

**Pull request closed without merging**